### PR TITLE
Add type check to functions/connection/*

### PR DESCRIPTION
**Pull request by soramichi**
Continued from #2255.

Ths PR adds type checking codes to `functions/connection/*.py` to see if numpy and cupy are mixed (and outputs a graceful warning when they are).

I confirmed that all the related tests with @attr.gpu pass in my local environment.

### CODE REVIEW

File: chainer/functions/connection/bilinear.py [hunk 1]
```diff
@@ -51,6 +51,11 @@ def forward(self, inputs):
         e2 = array.as_mat(inputs[1])
         W = inputs[2]
 
+        if(not type_check.same_types(*inputs)):
+            raise ValueError('numpy and cupy must not be used together\n'
+                             'type(W): %s, type(e1): %s, type(e2): %s'
+                             % (type(W), type(e1), type(e2)))
```
Comment (niboshi) [inline chainer/functions/connection/bilinear.py]: Formatting with `%` operator is discouraged. Could you use `str.format()` instead?

**Comment by soramichi** [comment]
Hi, thank you for the review.

I fixed the coding style as suggested, applied the same fix to `embed_id.py` (which had been already merged in #2255), and inserted the type checking codes to `dilated_convolution_2d.py` that I missed previously.

File: chainer/functions/connection/bilinear.py [hunk 2]
```diff
@@ -77,6 +82,12 @@ def backward(self, inputs, grad_outputs):
         e1 = array.as_mat(inputs[0])
         e2 = array.as_mat(inputs[1])
         W = inputs[2]
+
+        if(not type_check.same_types(*inputs)):
+            raise ValueError('numpy and cupy must not be used together\n'
+                             'type(W): %s, type(e1): %s, type(e2): %s'
+                             % (type(W), type(e1), type(e2)))
+
```
Comment (niboshi) [inline chainer/functions/connection/bilinear.py]: Please fix this section as well.

**Comment by soramichi** [comment]
oops, sorry for the mistake. fixed.

**Comment by niboshi** [comment]
LGTM. Thank you for your contribution!

**Pull request merged**